File: pkg/client/signer.go
package client

import (
	"context"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/big"
	"strings"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ivanzzeth/ethsig"
	"github.com/ivanzzeth/ethsig/eip712"
)

// RemoteSigner implements ethsig signer interfaces by making remote calls to the signing service.
// It can be used as a drop-in replacement for local signers.
type RemoteSigner struct {
	client  *Client
	address common.Address
	chainID string
}

// Ensure RemoteSigner implements all ethsig interfaces.
var (
	_ ethsig.AddressGetter     = (*RemoteSigner)(nil)
	_ ethsig.RawMessageSigner  = (*RemoteSigner)(nil)
	_ ethsig.HashSigner        = (*RemoteSigner)(nil)
	_ ethsig.EIP191Signer      = (*RemoteSigner)(nil)
	_ ethsig.PersonalSigner    = (*RemoteSigner)(nil)
	_ ethsig.TypedDataSigner   = (*RemoteSigner)(nil)
	_ ethsig.TransactionSigner = (*RemoteSigner)(nil)
)

// GetSigner returns a RemoteSigner for the specified address.
// The returned signer implements all ethsig signer interfaces.
func (c *Client) GetSigner(address common.Address, chainID string) *RemoteSigner {
	return &RemoteSigner{
		client:  c,
		address: address,
		chainID: chainID,
	}
}

// GetAddress returns the signer's address.
// Implements ethsig.AddressGetter.
func (s *RemoteSigner) GetAddress() common.Address {
	return s.address
}

// SignRawMessage signs raw message bytes.
// Implements ethsig.RawMessageSigner.
func (s *RemoteSigner) SignRawMessage(raw []byte) ([]byte, error) {
	return s.SignRawMessageWithContext(context.Background(), raw)
}

// SignRawMessageWithContext signs raw message bytes with context.
func (s *RemoteSigner) SignRawMessageWithContext(ctx context.Context, raw []byte) ([]byte, error) {
	payload := RawMessagePayload{
		RawMessage: raw,
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	resp, err := s.client.Sign(ctx, &SignRequest{
		ChainID:       s.chainID,
		SignerAddress: s.address.Hex(),
		SignType:      SignTypeRawMessage,
		Payload:       payloadBytes,
	})
	if err != nil {
		return nil, err
	}

	return decodeSignature(resp.Signature)
}

// SignHash signs pre-hashed data (32 bytes).
// Implements ethsig.HashSigner.
func (s *RemoteSigner) SignHash(hashedData common.Hash) ([]byte, error) {
	return s.SignHashWithContext(context.Background(), hashedData)
}

// SignHashWithContext signs pre-hashed data with context.
func (s *RemoteSigner) SignHashWithContext(ctx context.Context, hashedData common.Hash) ([]byte, error) {
	payload := HashPayload{
		Hash: hashedData.Hex(),
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	resp, err := s.client.Sign(ctx, &SignRequest{
		ChainID:       s.chainID,
		SignerAddress: s.address.Hex(),
		SignType:      SignTypeHash,
		Payload:       payloadBytes,
	})
	if err != nil {
		return nil, err
	}

	return decodeSignature(resp.Signature)
}

// SignEIP191Message signs an EIP-191 formatted message.
// Implements ethsig.EIP191Signer.
func (s *RemoteSigner) SignEIP191Message(message string) ([]byte, error) {
	return s.SignEIP191MessageWithContext(context.Background(), message)
}

// SignEIP191MessageWithContext signs an EIP-191 formatted message with context.
func (s *RemoteSigner) SignEIP191MessageWithContext(ctx context.Context, message string) ([]byte, error) {
	payload := MessagePayload{
		Message: message,
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	resp, err := s.client.Sign(ctx, &SignRequest{
		ChainID:       s.chainID,
		SignerAddress: s.address.Hex(),
		SignType:      SignTypeEIP191,
		Payload:       payloadBytes,
	})
	if err != nil {
		return nil, err
	}

	return decodeSignature(resp.Signature)
}

// PersonalSign signs data using personal_sign (EIP-191 0x45).
// Implements ethsig.PersonalSigner.
func (s *RemoteSigner) PersonalSign(data string) ([]byte, error) {
	return s.PersonalSignWithContext(context.Background(), data)
}

// PersonalSignWithContext signs data using personal_sign with context.
func (s *RemoteSigner) PersonalSignWithContext(ctx context.Context, data string) ([]byte, error) {
	payload := MessagePayload{
		Message: data,
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	resp, err := s.client.Sign(ctx, &SignRequest{
		ChainID:       s.chainID,
		SignerAddress: s.address.Hex(),
		SignType:      SignTypePersonal,
		Payload:       payloadBytes,
	})
	if err != nil {
		return nil, err
	}

	return decodeSignature(resp.Signature)
}

// SignTypedData signs EIP-712 typed data.
// Implements ethsig.TypedDataSigner.
func (s *RemoteSigner) SignTypedData(typedData eip712.TypedData) ([]byte, error) {
	return s.SignTypedDataWithContext(context.Background(), typedData)
}

// SignTypedDataWithContext signs EIP-712 typed data with context.
func (s *RemoteSigner) SignTypedDataWithContext(ctx context.Context, typedData eip712.TypedData) ([]byte, error) {
	// Convert ethsig types to client types
	clientTypedData := convertToClientTypedData(typedData)

	payload := TypedDataPayload{
		TypedData: clientTypedData,
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	resp, err := s.client.Sign(ctx, &SignRequest{
		ChainID:       s.chainID,
		SignerAddress: s.address.Hex(),
		SignType:      SignTypeTypedData,
		Payload:       payloadBytes,
	})
	if err != nil {
		return nil, err
	}

	return decodeSignature(resp.Signature)
}

// SignTransactionWithChainID signs an Ethereum transaction with explicit chain ID.
// Implements ethsig.TransactionSigner.
func (s *RemoteSigner) SignTransactionWithChainID(tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
	return s.SignTransactionWithChainIDAndContext(context.Background(), tx, chainID)
}

// SignTransactionWithChainIDAndContext signs an Ethereum transaction with context.
func (s *RemoteSigner) SignTransactionWithChainIDAndContext(ctx context.Context, tx *types.Transaction, chainID *big.Int) (*types.Transaction, error) {
	// Determine transaction type
	txType := "legacy"
	switch tx.Type() {
	case types.AccessListTxType:
		txType = "eip2930"
	case types.DynamicFeeTxType:
		txType = "eip1559"
	}

	// Build transaction payload
	txPayload := &Transaction{
		Value:  tx.Value().String(),
		Gas:    tx.Gas(),
		TxType: txType,
	}

	if tx.To() != nil {
		to := tx.To().Hex()
		txPayload.To = &to
	}

	if len(tx.Data()) > 0 {
		txPayload.Data = "0x" + hex.EncodeToString(tx.Data())
	}

	nonce := tx.Nonce()
	txPayload.Nonce = &nonce

	// Set gas price fields based on tx type
	switch txType {
	case "legacy":
		txPayload.GasPrice = tx.GasPrice().String()
	case "eip1559":
		txPayload.GasTipCap = tx.GasTipCap().String()
		txPayload.GasFeeCap = tx.GasFeeCap().String()
	case "eip2930":
		txPayload.GasPrice = tx.GasPrice().String()
	}

	payload := TransactionPayload{
		Transaction: txPayload,
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	// Use the provided chainID
	reqChainID := s.chainID
	if chainID != nil {
		reqChainID = chainID.String()
	}

	resp, err := s.client.Sign(ctx, &SignRequest{
		ChainID:       reqChainID,
		SignerAddress: s.address.Hex(),
		SignType:      SignTypeTransaction,
		Payload:       payloadBytes,
	})
	if err != nil {
		return nil, err
	}

	// Decode the signed transaction from the response
	if resp.SignedData == "" {
		return nil, fmt.Errorf("no signed transaction data in response")
	}

	signedTxBytes, err := decodeHexOrBase64(resp.SignedData)
	if err != nil {
		return nil, fmt.Errorf("failed to decode signed transaction: %w", err)
	}

	var signedTx types.Transaction
	if err := signedTx.UnmarshalBinary(signedTxBytes); err != nil {
		return nil, fmt.Errorf("failed to unmarshal signed transaction: %w", err)
	}

	return &signedTx, nil
}

// Close is a no-op for RemoteSigner as there's no sensitive local data to clean up.
func (s *RemoteSigner) Close() error {
	return nil
}

// convertToClientTypedData converts ethsig TypedData to client TypedData.
func convertToClientTypedData(td eip712.TypedData) *TypedData {
	// Convert types
	clientTypes := make(map[string][]TypedDataField)
	for name, fields := range td.Types {
		clientFields := make([]TypedDataField, len(fields))
		for i, f := range fields {
			clientFields[i] = TypedDataField{
				Name: f.Name,
				Type: f.Type,
			}
		}
		clientTypes[name] = clientFields
	}

	// Convert domain
	clientDomain := TypedDataDomain{
		Name:              td.Domain.Name,
		Version:           td.Domain.Version,
		ChainId:           td.Domain.ChainId,
		VerifyingContract: td.Domain.VerifyingContract,
		Salt:              td.Domain.Salt,
	}

	return &TypedData{
		Types:       clientTypes,
		PrimaryType: td.PrimaryType,
		Domain:      clientDomain,
		Message:     td.Message,
	}
}

// decodeSignature decodes a signature from hex or base64 format.
func decodeSignature(sig string) ([]byte, error) {
	if sig == "" {
		return nil, fmt.Errorf("empty signature")
	}

	return decodeHexOrBase64(sig)
}

// decodeHexOrBase64 decodes a string that may be hex or base64 encoded.
func decodeHexOrBase64(s string) ([]byte, error) {
	// Try hex first (with 0x prefix)
	if strings.HasPrefix(s, "0x") {
		return hex.DecodeString(s[2:])
	}

	// Try hex without prefix
	if isHex(s) {
		return hex.DecodeString(s)
	}

	// Try base64
	return base64.StdEncoding.DecodeString(s)
}

// isHex checks if a string contains only hex characters.
func isHex(s string) bool {
	for _, c := range s {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
			return false
		}
	}
	return len(s) > 0 && len(s)%2 == 0
}

File: pkg/rs-client/src/signer.rs
//! Signer abstractions.
//!
//! These traits mirror the `ethsig` interfaces the Go SDK implements, so a
//! signing backend is a swappable dependency rather than a hard-wired one.
//! [`RemoteSigner`] is the implementation backed by the remote-signer service;
//! a caller can equally supply a local keystore, an HSM/KMS client or a test
//! double without the calling code changing.
//!
//! Every trait is object-safe, so `Box<dyn TransactionSigner>` works and the
//! backend can be chosen at runtime from configuration.

use crate::error::Error;
use crate::evm::{
    HashPayload, MessagePayload, RawMessagePayload, SignRequest, Transaction, TransactionPayload,
    TypedDataPayload, SIGN_TYPE_EIP191, SIGN_TYPE_HASH, SIGN_TYPE_PERSONAL, SIGN_TYPE_RAW_MESSAGE,
    SIGN_TYPE_TRANSACTION, SIGN_TYPE_TYPED_DATA,
};

/// Exposes the address a signer signs for.
pub trait AddressGetter {
    fn address(&self) -> &str;
}

/// Signs 32 pre-hashed bytes. No prefixing is applied.
pub trait HashSigner {
    fn sign_hash(&self, hash: &str) -> Result<Vec<u8>, Error>;
}

/// Signs raw bytes with no EIP-191 prefix.
pub trait RawMessageSigner {
    fn sign_raw_message(&self, raw: &[u8]) -> Result<Vec<u8>, Error>;
}

/// Signs an EIP-191 formatted message.
pub trait Eip191Signer {
    fn sign_eip191_message(&self, message: &str) -> Result<Vec<u8>, Error>;
}

/// Signs via `personal_sign` (EIP-191 version byte `0x45`).
pub trait PersonalSigner {
    fn personal_sign(&self, data: &str) -> Result<Vec<u8>, Error>;
}

/// Signs EIP-712 typed data.
pub trait TypedDataSigner {
    fn sign_typed_data(&self, typed_data: &serde_json::Value) -> Result<Vec<u8>, Error>;
}

/// Signs a transaction, returning the encoded signed transaction bytes ready
/// to broadcast.
pub trait TransactionSigner {
    fn sign_transaction(&self, tx: &Transaction) -> Result<Vec<u8>, Error>;
}

/// Accepts `0x`-prefixed hex, bare hex, or base64 — matching the Go SDK's
/// tolerance for how the server encodes signatures.
pub(crate) fn decode_hex_or_base64(s: &str) -> Result<Vec<u8>, Error> {
    use base64::engine::general_purpose::STANDARD as BASE64_STD;
    use base64::Engine;

    if let Some(rest) = s.strip_prefix("0x") {
        return hex::decode(rest)
            .map_err(|e| Error::InvalidConfig(format!("invalid hex signature: {e}")));
    }
    if !s.is_empty() && s.chars().all(|c| c.is_ascii_hexdigit()) {
        return hex::decode(s)
            .map_err(|e| Error::InvalidConfig(format!("invalid hex signature: {e}")));
    }
    BASE64_STD
        .decode(s)
        .map_err(|e| Error::InvalidConfig(format!("invalid base64 signature: {e}")))
}

pub(crate) fn decode_signature(sig: Option<&str>) -> Result<Vec<u8>, Error> {
    match sig {
        Some(s) if !s.is_empty() => decode_hex_or_base64(s),
        _ => Err(Error::InvalidPayload),
    }
}

pub(crate) fn build_request(
    chain_id: &str,
    address: &str,
    sign_type: &str,
    payload: serde_json::Value,
) -> SignRequest {
    SignRequest {
        chain_id: chain_id.to_string(),
        signer_address: address.to_string(),
        sign_type: sign_type.to_string(),
        payload,
    }
}

/// Builds the payload for each sign type. Shared by the blocking and async
/// signers so the wire format cannot diverge between them.
pub(crate) mod payload {
    use super::*;

    pub fn hash(hash: &str) -> Result<serde_json::Value, Error> {
        Ok(serde_json::to_value(HashPayload {
            hash: hash.to_string(),
        })?)
    }

    pub fn raw_message(raw: &[u8]) -> Result<serde_json::Value, Error> {
        Ok(serde_json::to_value(RawMessagePayload {
            raw_message: raw.to_vec(),
        })?)
    }

    pub fn message(message: &str) -> Result<serde_json::Value, Error> {
        Ok(serde_json::to_value(MessagePayload {
            message: message.to_string(),
        })?)
    }

    pub fn typed_data(typed_data: &serde_json::Value) -> Result<serde_json::Value, Error> {
        Ok(serde_json::to_value(TypedDataPayload {
            typed_data: typed_data.clone(),
        })?)
    }

    pub fn transaction(tx: &Transaction) -> Result<serde_json::Value, Error> {
        Ok(serde_json::to_value(TransactionPayload {
            transaction: tx.clone(),
        })?)
    }
}

/// A signer backed by the remote-signer service.
///
/// Submits through [`crate::evm::SignService`], so the service's rule engine
/// and budgets apply to every signature.
#[derive(Clone)]
pub struct RemoteSigner {
    sign: crate::evm::SignService,
    address: String,
    chain_id: String,
}

impl RemoteSigner {
    pub fn new(
        sign: crate::evm::SignService,
        address: impl Into<String>,
        chain_id: impl Into<String>,
    ) -> Self {
        Self {
            sign,
            address: address.into(),
            chain_id: chain_id.into(),
        }
    }

    pub fn chain_id(&self) -> &str {
        &self.chain_id
    }

    pub fn set_chain_id(&mut self, chain_id: impl Into<String>) {
        self.chain_id = chain_id.into();
    }

    fn submit(&self, sign_type: &str, payload: serde_json::Value) -> Result<Vec<u8>, Error> {
        let req = build_request(&self.chain_id, &self.address, sign_type, payload);
        let resp = self.sign.execute(&req)?;
        decode_signature(resp.signature.as_deref())
    }
}

impl AddressGetter for RemoteSigner {
    fn address(&self) -> &str {
        &self.address
    }
}

impl HashSigner for RemoteSigner {
    fn sign_hash(&self, hash: &str) -> Result<Vec<u8>, Error> {
        self.submit(SIGN_TYPE_HASH, payload::hash(hash)?)
    }
}

impl RawMessageSigner for RemoteSigner {
    fn sign_raw_message(&self, raw: &[u8]) -> Result<Vec<u8>, Error> {
        self.submit(SIGN_TYPE_RAW_MESSAGE, payload::raw_message(raw)?)
    }
}

impl Eip191Signer for RemoteSigner {
    fn sign_eip191_message(&self, message: &str) -> Result<Vec<u8>, Error> {
        self.submit(SIGN_TYPE_EIP191, payload::message(message)?)
    }
}

impl PersonalSigner for RemoteSigner {
    fn personal_sign(&self, data: &str) -> Result<Vec<u8>, Error> {
        self.submit(SIGN_TYPE_PERSONAL, payload::message(data)?)
    }
}

impl TypedDataSigner for RemoteSigner {
    fn sign_typed_data(&self, typed_data: &serde_json::Value) -> Result<Vec<u8>, Error> {
        self.submit(SIGN_TYPE_TYPED_DATA, payload::typed_data(typed_data)?)
    }
}

impl TransactionSigner for RemoteSigner {
    fn sign_transaction(&self, tx: &Transaction) -> Result<Vec<u8>, Error> {
        let req = build_request(
            &self.chain_id,
            &self.address,
            SIGN_TYPE_TRANSACTION,
            payload::transaction(tx)?,
        );
        let resp = self.sign.execute(&req)?;
        match resp.signed_data.as_deref() {
            Some(s) if !s.is_empty() => decode_hex_or_base64(s),
            _ => Err(Error::InvalidPayload),
        }
    }
}

#[cfg(feature = "async")]
mod asynchronous {
    use super::*;
    use async_trait::async_trait;

    #[async_trait]
    pub trait AsyncHashSigner: Send + Sync {
        async fn sign_hash(&self, hash: &str) -> Result<Vec<u8>, Error>;
    }

    #[async_trait]
    pub trait AsyncRawMessageSigner: Send + Sync {
        async fn sign_raw_message(&self, raw: &[u8]) -> Result<Vec<u8>, Error>;
    }

    #[async_trait]
    pub trait AsyncEip191Signer: Send + Sync {
        async fn sign_eip191_message(&self, message: &str) -> Result<Vec<u8>, Error>;
    }

    #[async_trait]
    pub trait AsyncPersonalSigner: Send + Sync {
        async fn personal_sign(&self, data: &str) -> Result<Vec<u8>, Error>;
    }

    #[async_trait]
    pub trait AsyncTypedDataSigner: Send + Sync {
        async fn sign_typed_data(&self, typed_data: &serde_json::Value)
            -> Result<Vec<u8>, Error>;
    }

    #[async_trait]
    pub trait AsyncTransactionSigner: Send + Sync {
        async fn sign_transaction(&self, tx: &Transaction) -> Result<Vec<u8>, Error>;
    }

    /// Non-blocking counterpart of [`RemoteSigner`].
    #[derive(Clone)]
    pub struct AsyncRemoteSigner {
        sign: crate::evm::AsyncSignService,
        address: String,
        chain_id: String,
    }

    impl AsyncRemoteSigner {
        pub fn new(
            sign: crate::evm::AsyncSignService,
            address: impl Into<String>,
            chain_id: impl Into<String>,
        ) -> Self {
            Self {
                sign,
                address: address.into(),
                chain_id: chain_id.into(),
            }
        }

        pub fn chain_id(&self) -> &str {
            &self.chain_id
        }

        pub fn set_chain_id(&mut self, chain_id: impl Into<String>) {
            self.chain_id = chain_id.into();
        }

        /// Automated callers should not sit on a human approval; this uses the
        /// non-waiting submit so a rule that requires approval surfaces at once.
        async fn submit(
            &self,
            sign_type: &str,
            payload: serde_json::Value,
        ) -> Result<Vec<u8>, Error> {
            let req = build_request(&self.chain_id, &self.address, sign_type, payload);
            let resp = self.sign.execute_no_wait(&req).await?;
            decode_signature(resp.signature.as_deref())
        }
    }

    impl AddressGetter for AsyncRemoteSigner {
        fn address(&self) -> &str {
            &self.address
        }
    }

    #[async_trait]
    impl AsyncHashSigner for AsyncRemoteSigner {
        async fn sign_hash(&self, hash: &str) -> Result<Vec<u8>, Error> {
            self.submit(SIGN_TYPE_HASH, payload::hash(hash)?).await
        }
    }

    #[async_trait]
    impl AsyncRawMessageSigner for AsyncRemoteSigner {
        async fn sign_raw_message(&self, raw: &[u8]) -> Result<Vec<u8>, Error> {
            self.submit(SIGN_TYPE_RAW_MESSAGE, payload::raw_message(raw)?)
                .await
        }
    }

    #[async_trait]
    impl AsyncEip191Signer for AsyncRemoteSigner {
        async fn sign_eip191_message(&self, message: &str) -> Result<Vec<u8>, Error> {
            self.submit(SIGN_TYPE_EIP191, payload::message(message)?)
                .await
        }
    }

    #[async_trait]
    impl AsyncPersonalSigner for AsyncRemoteSigner {
        async fn personal_sign(&self, data: &str) -> Result<Vec<u8>, Error> {
            self.submit(SIGN_TYPE_PERSONAL, payload::message(data)?)
                .await
        }
    }

    #[async_trait]
    impl AsyncTypedDataSigner for AsyncRemoteSigner {
        async fn sign_typed_data(
            &self,
            typed_data: &serde_json::Value,
        ) -> Result<Vec<u8>, Error> {
            self.submit(SIGN_TYPE_TYPED_DATA, payload::typed_data(typed_data)?)
                .await
        }
    }

    #[async_trait]
    impl AsyncTransactionSigner for AsyncRemoteSigner {
        async fn sign_transaction(&self, tx: &Transaction) -> Result<Vec<u8>, Error> {
            let req = build_request(
                &self.chain_id,
                &self.address,
                SIGN_TYPE_TRANSACTION,
                payload::transaction(tx)?,
            );
            let resp = self.sign.execute_no_wait(&req).await?;
            match resp.signed_data.as_deref() {
                Some(s) if !s.is_empty() => decode_hex_or_base64(s),
                _ => Err(Error::InvalidPayload),
            }
        }
    }
}

#[cfg(feature = "async")]
pub use asynchronous::{
    AsyncEip191Signer, AsyncHashSigner, AsyncPersonalSigner, AsyncRawMessageSigner,
    AsyncRemoteSigner, AsyncTransactionSigner, AsyncTypedDataSigner,
};

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn signature_decoding_accepts_hex_and_base64() {
        assert_eq!(decode_hex_or_base64("0x010203").unwrap(), vec![1, 2, 3]);
        assert_eq!(decode_hex_or_base64("010203").unwrap(), vec![1, 2, 3]);
        assert_eq!(decode_hex_or_base64("AQID").unwrap(), vec![1, 2, 3]);
    }

    #[test]
    fn empty_or_missing_signature_is_an_error() {
        assert!(matches!(
            decode_signature(None),
            Err(Error::InvalidPayload)
        ));
        assert!(matches!(
            decode_signature(Some("")),
            Err(Error::InvalidPayload)
        ));
    }

    #[test]
    fn payloads_are_wrapped_the_way_the_server_expects() {
        assert_eq!(payload::hash("0xabc").unwrap()["hash"], "0xabc");
        assert_eq!(payload::message("hi").unwrap()["message"], "hi");
        assert_eq!(payload::raw_message(&[1, 2, 3]).unwrap()["raw_message"], "AQID");

        let tx = Transaction::legacy("0", 21000, "1").nonce(3);
        let p = payload::transaction(&tx).unwrap();
        assert_eq!(p["transaction"]["nonce"], 3);

        let td = serde_json::json!({"primaryType": "Mail"});
        assert_eq!(payload::typed_data(&td).unwrap()["typed_data"]["primaryType"], "Mail");
    }

    /// A second backend can satisfy the same traits — that is the point of
    /// having them.
    struct FixedSigner(Vec<u8>);

    impl AddressGetter for FixedSigner {
        fn address(&self) -> &str {
            "0xfixed"
        }
    }

    impl TransactionSigner for FixedSigner {
        fn sign_transaction(&self, _tx: &Transaction) -> Result<Vec<u8>, Error> {
            Ok(self.0.clone())
        }
    }

    #[test]
    fn traits_are_object_safe_so_backends_are_swappable() {
        let signer: Box<dyn TransactionSigner> = Box::new(FixedSigner(vec![9, 9]));
        let tx = Transaction::legacy("0", 21000, "1").nonce(0);
        assert_eq!(signer.sign_transaction(&tx).unwrap(), vec![9, 9]);
    }
}
